cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/SparkStreamingPipelineDriver.java

```diff
@@ -248,4 +285,46 @@ private JavaStreamingContext run(DataStreamsPipelineSpec pipelineSpec,
   private boolean ensureDirExists(FileSystem fileSystem, Path dir) throws IOException {
     return fileSystem.isDirectory(dir) || fileSystem.mkdirs(dir) || fileSystem.isDirectory(dir);
   }
+
+  /**
+   * Reports if all data blocks stored in the checkpoint generated by receivers are processed.
+   *
+   * @param jssc the spark streaming context for the streaming process
+   * @param pipelineSpec the specification of the pipeline
+   * @param checkpointDir the checkpoint directory configured for the pipeline
+   * @return {@code true} if all data blocks are processed; {@code false} otherwise
+   */
+  private boolean hasAllBlocksProcessed(JavaStreamingContext jssc,
+                                        DataStreamsPipelineSpec pipelineSpec,
+                                        String checkpointDir) {
+    try {
+      Option<Checkpoint> checkpointOption = CheckpointReader.read(checkpointDir);
+
+      // If there is no checkpoint file, it could be no micro-batch job was executed at all.
+      // To keep it safe, assuming there is unprocessed data.
+      if (checkpointOption.isEmpty()) {
+        return false;
```

Comment:
If the receiver is reading from an exhausted source, the microbatches will always be empty, right? Doesn't this become a problem if graceful shutdown is invoked for an exhausted source?

Reply:
Since checkpoint is enabled for the driver, it doesn't matter where there is receiver involved or not, there should be some checkpoint files there. If the checkpoint directory is really empty during shutdown, it doesn't matter as there is nothing to delete as well.